### PR TITLE
add Half support for flash attention

### PR DESCRIPTION
**Pull request by CaoE**
Re-open for https://github.com/pytorch/pytorch/pull/118368.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by CaoE** [comment]
@clee2000 Could you please help test this PR in internal builds ?

**Comment by drisspg** [comment]
@larryliu0820  I think that you have failing tests on this PR: https://github.com/pytorch/pytorch/pull/119239 because it was originally based off of the landed version of this. I forsee a merge conflict, so it is probably better to stage this so that your's lands @larryliu0820 with an `xfail` on the float16 case. And then this can be based ontop of that PR and can remove the xfail

**Comment by larryliu0820** [comment]
> @larryliu0820 I think that you have failing tests on this PR: #119239 because it was originally based off of the landed version of this. I forsee a merge conflict, so it is probably better to stage this so that your's lands @larryliu0820 with an `xfail` on the float16 case. And then this can be based ontop of that PR and can remove the xfail

Ok so sounds like I should rebase my PR to main?

**Comment by mikekgfb** [comment]
Does #119247 include #119082 as part of the updated PR, or should it be landed concurrently with that PR.

**Comment by larryliu0820** [comment]
@CaoE Hello I just merged https://github.com/pytorch/pytorch/pull/119239 and if you rebase to main it's likely you'll see the test failure on `test/test_decomp.py`.  If that happens please fix `decompositions.py` accordingly. Thanks!

**Review by drisspg — APPROVED** [review]

File: aten/src/ATen/native/transformers/attention.cpp [hunk 1]
```diff
@@ -760,8 +760,8 @@ _scaled_dot_product_flash_attention_cpu(
   int64_t num_head = query.size(1);
   int64_t headSize = query.size(3);
 
-  TORCH_CHECK(c10::isFloatingType(dtype) && dtype != ScalarType::Half,
-    "scaled_dot_product_attention_flash_attention: Expected data type in FP32, FP64, BF16, but got ", dtype, " instead.");
+  TORCH_CHECK(c10::isFloatingType(dtype),
+    "scaled_dot_product_attention_flash_attention: Expected data type in FP32, FP64, BF16, FP16, but got ", dtype, " instead.");
```
Comment (malfet) [inline aten/src/ATen/native/transformers/attention.cpp:764]: Nit
```suggestion
    "scaled_dot_product_attention_flash_attention: Expected floating data type, but got ", dtype, " instead.");
```

File: aten/src/ATen/native/transformers/attention.cpp [hunk 2]
```diff
@@ -760,8 +760,8 @@ _scaled_dot_product_flash_attention_cpu(
   int64_t num_head = query.size(1);
   int64_t headSize = query.size(3);
 
-  TORCH_CHECK(c10::isFloatingType(dtype) && dtype != ScalarType::Half,
-    "scaled_dot_product_attention_flash_attention: Expected data type in FP32, FP64, BF16, but got ", dtype, " instead.");
+  TORCH_CHECK(c10::isFloatingType(dtype),
```
Comment (malfet) [inline aten/src/ATen/native/transformers/attention.cpp:763]: Q: is FP8 considered floating type?

File: test/inductor/test_fused_attention.py [hunk 3]
```diff
@@ -118,14 +118,16 @@ def dot_prod_attention(
...
 
         for dtype in [torch.float, torch.half]:
-            if self.device == "cpu" and dtype == torch.half:
-                continue
+            atol = 0.001
             rtol = 1.3e-6 if dtype == torch.float else 0.7
-            self._check_common(dot_prod_attention, dtype=dtype, atol=0.001, rtol=rtol)
+            if self.device == "cpu" and dtype == torch.half:
+                atol = 2e-3
+                rtol = 1e-2
```
Comment (malfet) [inline test/inductor/test_fused_attention.py:125]: Please leave a comment why tolerance needs to be raised for CPU, but not for CUDA

File: test/test_transformers.py [hunk 4]
```diff
@@ -1754,14 +1754,14 @@ def test_fused_sdp_choice_cpu(self, device, type: str, dropout: float, dtype: to
         q, k, v = make_tensor(size), make_tensor(size), make_tensor(size)
         if type == "nested" \
                 or dropout > 0.0 \
-                or dtype not in [torch.float32, torch.float64, torch.bfloat16]:
+                or dtype not in [torch.float32, torch.float64, torch.bfloat16, torch.float16]:
```
Comment (malfet) [inline test/test_transformers.py:1757]: Nit
```suggestion
                or not dtype.is_floating_point:
```

File: test/test_transformers.py [hunk 5]
```diff
@@ -1754,14 +1754,14 @@ def test_fused_sdp_choice_cpu(self, device, type: str, dropout: float, dtype: to
...
             assert torch._fused_sdp_choice(q, k, v, dropout_p=dropout) == SDPBackend.MATH.value
         else:
             assert torch._fused_sdp_choice(q, k, v, dropout_p=dropout) == SDPBackend.FLASH_ATTENTION.value
 
     @onlyCPU
     @parametrize("fused_kernel", [SDPBackend.FLASH_ATTENTION])
-    @parametrize("dtype", [torch.float64, torch.float32, torch.bfloat16])
+    @parametrize("dtype", [torch.float64, torch.float32, torch.bfloat16, torch.float16])
```
Comment (malfet) [inline test/test_transformers.py:1764]: Nit
```suggestion
    @parametrize("dtype", get_all_fp_dtypes())
```

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/transformers/attention.cpp (hunk 2 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/attention.cpp:763]: ahh this is a good point it seems that c10::isFloatingtype also returns true for fp8 types we should ban those, I didn't realize the set isFloatingType included  those, good catch!

File: test/inductor/test_fused_attention.py (hunk 3 above)
Comment (CaoE) [reply test/inductor/test_fused_attention.py:125]: Sorry for the late reply, as I am currently on holiday. `rtol` on CPU does not need a value as large as 0.7.

**Comment by mikekgfb** [comment]
Please submit a PR when you return!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Pull request closed without merging**